src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs

```diff
@@ -77,30 +78,27 @@ public static class EnumerableMappingBuilder
         if (!ctx.IsConversionEnabled(MappingConversionType.Enumerable))
             return null;
 
-        if (BuildElementMapping(ctx) is not { } elementMapping)
+        if (ctx.CollectionInfos == null)
+            return null;
+
+        var elementMapping = ctx.FindOrBuildMapping(ctx.CollectionInfos.Source.EnumeratedType, ctx.CollectionInfos.Target.EnumeratedType);
+        if (elementMapping == null)
             return null;
 
-        if (ctx.Target.ImplementsGeneric(ctx.Types.Get(typeof(Stack<>)), out _))
+        if (ctx.CollectionInfos.Target.Type == CollectionType.Stack)
             return CreateForEach(nameof(Stack<object>.Push));
 
-        if (ctx.Target.ImplementsGeneric(ctx.Types.Get(typeof(Queue<>)), out _))
+        if (ctx.CollectionInfos.Target.Type == CollectionType.Queue)
             return CreateForEach(nameof(Queue<object>.Enqueue));
 
         // create a foreach loop with add calls if source is not an array
-        // and void ICollection.Add(T) or bool ISet.Add(T) is implemented and not explicit
-        // ensures add is not called and immutable types
-        // ISet.Add(T) is explicitly needed as sets implement the ICollection.Add(T) explicit,
-        // and override the add method with new
-        var hasImplicitCollectionAdd = ctx.Target.HasImplicitGenericImplementation(ctx.Types.Get(typeof(ICollection<>)), AddMethodName);
-        var hasImplicitSetAdd = ctx.Target.HasImplicitGenericImplementation(ctx.Types.Get(typeof(ISet<>)), AddMethodName);
-        if (!ctx.Target.IsArrayType() && (hasImplicitCollectionAdd || hasImplicitSetAdd))
-        {
+        // and has an implicit add method
+        // the implicit check is an easy way to exclude for ex. immutable types.
+        if (ctx.CollectionInfos.Target.Type != CollectionType.Array && ctx.CollectionInfos.Target.HasImplicitCollectionAddMethod)
             return CreateForEach(AddMethodName);
-        }
 
-        // if a mapping could be created for an immutable collection
-        // we diagnostic when it is an existing target mapping
-        if (ResolveImmutableCollectMethod(ctx) != null)
+        // if the target is an immutable collection emit a diagnostic
```

Comment:
I think the code is pretty self explanatory here, comment can be removed (just describes what the code does)

Reply:
Removed.